### PR TITLE
Fix an issue in the programming exercise details page for instructors

### PR DESCRIPTION
**Pull request by krusche**
Hibernate in combination with Hazelcast randomly leads to Null Pointer Exceptions when fetching template or solution participations for programming exercises in the `canAccessParticiption(...)` method. This happens more often on multi node sever setups, probably because of caching issues related to Hazelcast.

I changed how the programming exercise details page for instructors is loaded, which should solve the issue. The page needs a larger redesign with respect to the number of REST calls, but the change should solve the issue that you typically notice on Staging and Production, but not necessarily on the simple test servers (TS1, TS2, TS3, TS5) because they only have one server node.

I also removed a too similar method and unified the approach for `canAccessParticiption(...)`

### Checklist
- [x] I tested *all* changes and *all* related features with different users (student, tutor, editor, instructor, admin) on the test server https://artemistest.ase.in.tum.de.
- [x] Server: I added server integration tests 
- [x] Server: I followed the [coding and design guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/server/).
- [x] Server: I implemented the changes with a good performance and prevented too many database calls
- [x] Server: I documented the Java code using JavaDoc style.
- [x] Client: I followed the [coding and design guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/client/).
- [x] Client: I documented the TypeScript code using JSDoc style.

### Steps for Testing
0. Execute the following steps on the staging environment: https://artemis-staging.ase.in.tum.de
1. Open the programming exercise details page of a programming exercise and make sure the template and solution results (at the bottom) are displayed correctly
2. Open `Edit in Editor` and submit multiple times for template and solution, make sure the result changes
3. Regularly check if the programming exercise details page then displays the correct, i.e. most recent, result for template and solution without any error messages (after the build has taken place)

### CODE REVIEW

**Review by niclasschuemann — DISMISSED** [review]
**Observations:**
1. When creating a programming exercise and visiting the programming exercise detail view, the status of Template and Solution Result is "No result" instead of "Building and testing". Not sure if this is related though.
2. When editing the template in the online code editor, submitting and going back to the detail view displays the correct "Building and testing" state for the template submission. After the build is completed, the template submission result is updated correctly
3. The same as 2. applies to the solution participation
4. Submitting template and solution at the "same" time also results in a correct update of the Template and Submission Status on the programming exercise detail view
5. While testing the mentioned points, I got no errors in the network tab.

~**Note:** I did not look at the code yet.~
Code looks good to me, except for the missing bracket that Florian will mention in his review

File: src/main/java/de/tum/in/www1/artemis/domain/participation/StudentParticipation.java [hunk 1]
```diff
@@ -121,7 +121,7 @@ public boolean isOwnedBy(User user) {
     @Override
     public String toString() {
         String participantString = getStudent().map(student -> "student=" + student).orElse("team=" + team);
-        return "StudentParticipation{" + "id=" + getId() + ", presentationScore=" + presentationScore + ", " + participantString + "}";
+        return getClass().getSimpleName() + "" + "id=" + getId() + ", presentationScore=" + presentationScore + ", " + participantString + "}";
```
Comment (FlorianGlombik) [inline src/main/java/de/tum/in/www1/artemis/domain/participation/StudentParticipation.java:124]:
```suggestion
        return getClass().getSimpleName() + "{" + "id=" + getId() + ", presentationScore=" + presentationScore + ", " + participantString + "}";
```

**Review by FlorianGlombik — APPROVED** [review]
I have tested it on https://artemis-staging.ase.in.tum.de/ as artemis_test_user_1 and didn't encounter any issues, here is what I have tested (all modifications in the code editor):
- modify template
  - adding comment, submitting once
  - adding comment, submitting multiple times
  - passing test, updated properly
- solution
  - modified, submitting once
  - modified, submitting multiple times
- tests
  - modified, solution and template triggered
  
I have added a comment for a `toString()` method, I think a small mistake happened there.

The GitHub actions may also need an update, I am not quite sure about the unused import warnings. I have checked them locally and to me, it seems like these imports are used. We might want to deactivate these warnings if they don't function properly.

**Comment by krusche** [comment]
Thanks for reviewing, actually the import warnings were correct

**Pull request merged**